File: app/src/test/java/com/coinninja/coinkeeper/service/BroadcastTransactionServiceTest.java
package com.coinninja.coinkeeper.service;

import android.content.Intent;

import com.coinninja.coinkeeper.TestCoinKeeperApplication;
import com.coinninja.coinkeeper.model.UnspentTransactionHolder;
import com.coinninja.coinkeeper.model.db.PhoneNumber;
import com.coinninja.coinkeeper.model.dto.BroadcastTransactionDTO;
import com.coinninja.coinkeeper.model.dto.CompletedBroadcastDTO;
import com.coinninja.coinkeeper.service.client.model.Contact;
import com.coinninja.coinkeeper.service.runner.SaveTransactionRunner;
import com.coinninja.coinkeeper.util.Intents;
import com.coinninja.coinkeeper.util.PhoneNumberUtil;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InOrder;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.Robolectric;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import static org.mockito.Mockito.inOrder;
import static org.mockito.Mockito.mock;

@RunWith(RobolectricTestRunner.class)
@Config(application = TestCoinKeeperApplication.class)
public class BroadcastTransactionServiceTest {

    @Mock
    SaveTransactionRunner saveTransactionRunner;
    private BroadcastTransactionService broadcastTransactionService;
    private CompletedBroadcastDTO completedBroadcastActivityDTO;

    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.initMocks(this);
        broadcastTransactionService = Robolectric.setupService(BroadcastTransactionService.class);
        broadcastTransactionService.runner = saveTransactionRunner;
        BroadcastTransactionDTO broadcastActivityDTO = new BroadcastTransactionDTO(
                mock(UnspentTransactionHolder.class),
                new Contact(new PhoneNumber("+13333333333"), "Joe Blow", true),
                false, "--memo--",
                null);
        completedBroadcastActivityDTO = new CompletedBroadcastDTO(broadcastActivityDTO, "--txid--");
    }

    @Test
    public void saves_transaction_locally() {
        Intent intent = new Intent();
        intent.putExtra(Intents.EXTRA_COMPLETED_BROADCAST_DTO, completedBroadcastActivityDTO);

        broadcastTransactionService.onHandleIntent(intent);

        InOrder ordered = inOrder(saveTransactionRunner);
        ordered.verify(saveTransactionRunner).setCompletedBroadcastActivityDTO(completedBroadcastActivityDTO);
        ordered.verify(saveTransactionRunner).run();
    }
}

File: app/src/test/java/com/coinninja/coinkeeper/service/BroadcastTransactionServiceTest.kt
package com.coinninja.coinkeeper.service

import android.content.Intent
import androidx.test.ext.junit.runners.AndroidJUnit4
import app.coinninja.cn.libbitcoin.model.TransactionData
import com.coinninja.coinkeeper.model.Identity
import com.coinninja.coinkeeper.model.db.enums.IdentityType
import com.coinninja.coinkeeper.model.dto.BroadcastTransactionDTO
import com.coinninja.coinkeeper.model.dto.CompletedBroadcastDTO
import com.coinninja.coinkeeper.service.runner.SaveTransactionRunner
import com.coinninja.coinkeeper.util.DropbitIntents
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mockito.inOrder
import org.mockito.Mockito.mock
import org.robolectric.Robolectric

@RunWith(AndroidJUnit4::class)
class BroadcastTransactionServiceTest {

    private fun createDTO(): CompletedBroadcastDTO {
        val broadcastActivityDTO = BroadcastTransactionDTO(
                mock(TransactionData::class.java),
                false, "--memo--",
                Identity(IdentityType.PHONE, "+13305553333", "--hash--", "Joe Blow"),
                null)
        return CompletedBroadcastDTO(broadcastActivityDTO, "--txid--")
    }

    @Test
    fun saves_transaction_locally() {
        val service = Robolectric.setupService(BroadcastTransactionService::class.java)
        service.runner = mock(SaveTransactionRunner::class.java)
        val intent = Intent()
        val completedBroadcastDTO = createDTO()
        intent.putExtra(DropbitIntents.EXTRA_COMPLETED_BROADCAST_DTO, completedBroadcastDTO)

        service.onHandleIntent(intent)

        val ordered = inOrder(service.runner)
        ordered.verify(service.runner).setCompletedBroadcastActivityDTO(completedBroadcastDTO)
        ordered.verify(service.runner).run()
    }
}

